README.md
# featureservice

*Get all features from an Esri Feature Service*

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![Greenkeeper badge][greenkeeper-image]][greenkeeper-url]

[npm-image]: https://img.shields.io/npm/v/featureservice.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/featureservice
[travis-image]: https://img.shields.io/travis/koopjs/featureservice.svg?style=flat-square
[travis-url]: https://travis-ci.org/koopjs/featureservice
[greenkeeper-image]: https://badges.greenkeeper.io/koopjs/featureservice.svg
[greenkeeper-url]: https://greenkeeper.io/

A little module that extracts every feature from an Esri Feature Service. The real power in this module is that it's designed to page over a service and extract every single feature no matter what ArcGIS Server version the data is hosted on.

## Install

```
npm install featureservice
```

## Usage

```javascript
var FeatureService = require('featureservice')

// a url to a feature service
var url = 'http://....../FeatureServer/0'

var service = new FeatureService(url, options)
service.pages(function (err, pages) {
/* will give you links to all pages of data in the service*/
})
```

### Options
An object passed as the second parameter when initializing a service
- layer: the layer index to use
- size: the maximum page size when requesting features
- concurrency: the maximum concurrency for requesting features from a single server
- timeOut: the amount of time to wait with no response before cancelling a request
- logger: An object with a log method that takes a level and a message e.g. a Winston instance

### API

#### info(callback)
Get the information describing the service itself

#### layerIds(callback)
Get all the ids in a feature service layer

#### layerInfo(callback)
Get the json metadata for a service layer

#### statistics(field, stats, callback)
Get statistics for a field and an array of stats.

```javascript
service.statistics('id', ['min', 'max'], function (err, stats) {
  console.log(stats.features)
})
```

#### pages(callback)
Returns an array of page urls that would get every feature in the service

### Browser

A browser ready build of this module is in `dist/featureservice.min.js`.

#### Example

```html
<html>
  <script src="dist/featureservice.min.js"></script>
  <script>
    var service = new FeatureService('http://koop.dc.esri.com/socrata/seattle/2tje-83f6/FeatureServer/0', {})
    service.statistics('id', ['max'], function (err, stats) {
      console.log(err, stats)
    })
  </script>
</html>
```

## License

[Apache 2.0](LICENSE)

<!-- [](Esri Tags: ArcGIS Web Mapping GeoJson FeatureServices) -->
<!-- [](Esri Language: JavaScript) -->


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.6.3] - 2019-11-05
### Fixed
* browserify build

## [1.6.2] - 2019-11-05
### Changed
* Throw `TypeError` when URL is not a FeatureServer or MapServer with optional layer

## [1.6.1] - 2018-11-28
### Changed
* Double check pagination support with a request for the second page

## [1.6.0] - 2018-06-19
### Added
* Check for an option `outSR` and if provided use it to transform data; if no outSR option default to `outSR=4326`

## [1.5.13] - 2018-01-31
### Fixed
* Parse out html at the end of an otherwise valid response

## [1.5.12] - 2017-12-29
### Changed
* Z value is requested when paging

## [1.5.11] - 2016-10-23
### Fixed
* Use options.size to restrict max

## [1.5.10] - 2016-10-23
### Changed
* Don't use forever agent on http requests

## [1.5.9] - 2016-10-20
### Changed
* Use request library instead of hand-rolled http(s) requests

## [1.5.8] - 2016-10-16
### Changed
* Don't set min max OIDs for a single page request

## [1.5.7] - 2016-04-07
### Changed
* Remove unneeded try/catch

## [1.5.6] - 2016-04-06
### Fixed
* Get correct objectID range even when field names are unexpectedly capitalized

## [1.5.5] - 2016-03-31
### Changed
* Exposed `getObjectIdRange` as public function

### Fixed
* Corrected bug in statistics paging strategy

## [1.5.4] - 2016-03-30
### Fixed
* Don't throw exception when passed in layer is of type numeric

## [1.5.3] - 2016-03-23
### Changed
* A passed-in layer will always override what's in the url

## [1.5.2] - 2016-03-03
### Fixed
* Get the object id field if it's explicitly in info properties
* resultOffset strategy was reversed

## [1.5.1] - 2016-03-02
### Changed
* Don't use resultOffset for services not hosted on ArcGIS Online, it's not reliable
* Decode streaming server response

## [1.5.0] - 2016-02-28
### Added
* Service Info, Layer Info, and Metadata are memoized
* Can call `info`, `layerInfo`, and `metadata` without a callback when info is memoized

### Changed
* Concurrency is set during paging process
* URL parsing for services is more precise

## [1.4.6] - 2016-01-20
### Fixed
* Guard against missing ObjectID field when trying to page
* Remove unused dep

## [1.4.5] - 2016-01-13
### Changed
* Throttle concurrency down on every 2 failed requests
* Throttle concurrency up on every 10 successful requests

## [1.4.4] - 2016-01-07
### Fixed
* Removed bad reference to `this`

## [1.4.3] - 2016-01-04
### Changed
* Concurrency is throttled when encountering errors in the paging queue
### Fixed
* Urls and Layers are parsed correctly on service init

## [1.4.2] - 2015-10-13
### Changed
* Removed callback from parse try statement

## [1.4.1] - 2015-10-13
### Fixed
* Removed log statement that was causing the process to crash

## [1.4.0] - 2015-10-12
### Added
* Backoff time for failed requests can now be set as an option
* `service.info` returns information about the feature or map service

### Changed
* Return the page that was successfully completed when getting features
* Response parsing has been moved to its own function and the callback is no longer wrapped in a try/catch

### Fixed
* No longer failing to parse multi-chunk feature service responses that are not compressed

## [1.3.1]
### Fixed
* Log retry urls correclty

## [1.3.0]
### Added
* Allow a logger to be passed in through options.logger when initializing a new service

### Fixed
* Don't throw exception when there is no error in the json returned from server
* Catch non-json responses explicitly
* Don't use zlib methods unsupported in node 0.10.* in testing
* Error timestamps are all written in same case

## [1.2.7] - 2015-09-03
### Fixed
* Correct logic for correcting against poorly instantiated FeatureServices

## [1.2.6] - 2015-09-02
### Fixed
* Don't set a layer id with a query string
* Don't set a url with a query string

## [1.2.5] - 2015-09-02
### Fixed
* Uncaught error when `json.error` doesn't exist in `FeatureService.layerInfo`

## [1.2.4] - 2015-09-01
### Changed
* Restrict max record count to 5000 or less

### Fixed
* Pages generated from statistics use the correct layer when index > 0

## [1.2.3] - 2015-08-31
### Fixed
* Do not exclude object ids === 0 when building pages

## [1.2.2] - 2015-08-26
### Changed
* Errors after timeouts conform to standard
* Errors on metadata requests conform to standard

### Fixed
* Requests are ended correctly after timeouts

## [1.2.1] - 2015-08-17
### Changed
* All requests accept gzip or deflate compressed
* Requests are decoded async (for compatibility with node < 0.11.12)

### Fixed
* Pages are built correctly for layers with an index > 0

## [1.2.0] - 2015-08-11
### Fixed
* Pages based on object ids are now formed correctly

### Changed
* All errors returned to the requestor are standardized

## [1.1.1] - 2015-08-10
### Fixed
* Build dist for changes

## [1.1.0] - 2015-08-10
### Added
* New fixtures and integration tests for paging
* Support for paging layers from server version 10.0
* New fixtures and tests for decoding

### Changed
* Refactored paging strategy
* Moved feature request decoding into isolated function

### Fixed
* Catch errors that come on 200 responses
* Errors are reported correctly up the chain
* Retries for all errors

## [1.0.0] - 2015-08-07
### Added

lib/utils.js
module.exports = {
  /**
   * Given a service url and a geometry type, determines a default concurrency for requests
   *
   * @param {boolean} hosted - whether or not the service is hosted on ArcGIS Online
   * @param {string} geomType - the geometry type of the features in the service
   * @return {integer} the suggested concurrency
   */
  setConcurrency: function (hosted, geomType) {
    var naieve = hosted ? 16 : 4
    if (!geomType) return naieve
    var concurrency = geomType.match(/point/i) ? naieve : naieve / 4
    return Math.floor(concurrency)
  },
  /**
   * Parsed the layer and server from a feature service url
   * @param {string} url - a link to a feature service
   * @return {object} contains the layer, the server and whether or not the server is hosted
   */
  parseUrl: function (url) {
    // @todo: use the URL class once pre-node 6 has been deprecated
    var match = url.match(/^(.+?\/(?:feature|map)server)(?:\/(\d+))?/i)

    if (null === match) {
      throw new TypeError('unable to parse ' + url + ' as a mapserver or featureserver with optional layer')
    }

    return {
      layer: match[2],
      server: match[1],
      hosted: /services(\d)?(qa|dev)?.arcgis.com/.test(url)
    }
  },
  /**
   * Strip characters off a layer that don't belong
   * @param {string} raw - a raw layer options
   * @return {string} the layer index
   */
  sanitizeLayer: function (raw) {
    if (typeof raw === 'number') return raw
    else if (typeof raw !== 'string') return undefined

    var match = raw.match(/\/?(\d+)/)
    return match && match[1] ? match[1] : undefined
  }
}


index.js
var queue = require('async').queue
var Utils = require('./lib/utils.js')

/**
 * Feature Service constructor. Requires a URL.
 * Exposes a pageQueue that can take an array of page URLs.
 * The pageQueue will report back each page of features as they return.
 *
 * @class
 * @param {string} url - address of feature service
 * @param {object} options - layer (default: 0)
 */
var FeatureService = function (url, options) {
  // catch omission of `new` keyword
  if (!(this instanceof FeatureService)) {
    return new FeatureService(url, options)
  }
  var service = Utils.parseUrl(url)
  this.hosted = service.hosted
  this.server = service.server
  this.options = options || {}
  if (this.options.layer || this.options.layer === 0) this.options.layer = Utils.sanitizeLayer(this.options.layer)
  this.options.size = this.options.size || 5000
  this.options.backoff = this.options.backoff || 1000
  this.options.timeOut = this.options.timeOut || (1.5 * 60 * 1000)
  this.layer = this.options.layer || service.layer || 0
  this.outSR = this.options.outSR || 4326

  this.logger = this.options.logger

  this._request = require('xhr-request')

  // an async for requesting pages of data
  this.pageQueue = queue(this._requestFeatures.bind(this), this.options.concurrency || 4)
}

/**
 * Wraps logging functionality so a logger can be passed in
 *
 * @param {string} level - the log level to use
 * @param {string} message - the message to log
 */
FeatureService.prototype.log = function (level, message) {
  if (this.logger && this.logger.log) return this.logger.log(level, message)
  this._console(level, message)
}

/**
 * Wraps console logging to make it testable
 *
 * @param {string} level - the log level to use
 * @param {string} message - the message to log
 */
FeatureService.prototype._console = function (level, message) {
  switch (level) {
    case 'info':
      console.info(message)
      break
    case 'warn':
      console.warn(message)
      break
    case 'error':
      console.error(message)
      break
    default:
      console.log(message)
  }
}

/**
 * Wrap the request.get method for easier testing
 * @param {string} url
 * @param {function} callback
 */
// TODO combine this with _requestFeatures
FeatureService.prototype.request = function (url, callback) {
  var json
  // to ensure things are encoded just right for ArcGIS
  var encoded = encodeURI(decodeURI(url))
  var options = {
    timeout: this.options.timeOut,
    headers: {
      'user-agent': 'Featureservices-Node'
    }

  }
  this._request(encoded, options, function (err, data, res) {
    if (err) {
      if (err.code === 'ESOCKETTIMEDOUT') err.code = 504
      return callback(err)
    }
    try {
      json = JSON.parse(data)
    } catch (err) {
      // sometimes we get html or plain strings back
      var pattern = new RegExp(/[^{\[]/) // eslint-disable-line
      if (data.slice(0, 1).match(pattern)) {
        return callback(new Error('Received HTML or plain text when expecting JSON'))
      } else if (/\n<!DOCTYPE html>(\n|.|\s)*$/.test(data)) {
        try {
          json = JSON.parse(data.replace(/\n<!DOCTYPE html>(\n|.|\s)*$/, ''))
          return callback(null, json)
        } catch (e) {
          return callback(new Error('Failed to parse server response'))
        }
      }
      return callback(new Error('Failed to parse server response'))
    }
    callback(null, json)
  })
}

/**
 * Builds a url for querying the min/max values of the object id
 *
 * @param {string} field - the name of a field to build a stat request for
 * @returns {string} url
 */
FeatureService.prototype._statsUrl = function (field, stats) {
  field = field || this.options.objectIdField
  var json = []

  stats.forEach(function (stat) {
    json.push({
      'statisticType': stat,
      'onStatisticField': field,
      'outStatisticFieldName': stat + '_' + field
    })
  })

  return this.server + '/' + this.layer + '/query?f=json&outFields=&outStatistics=' + JSON.stringify(json)
}

/**
 * Gets the feature service info
 * @param {string} field - the name of a field to build a stat request for
 * @param {array} stats - an array of stats to request: ['min', 'max', 'avg', 'stddev', 'count']
 */
FeatureService.prototype.statistics = function (field, stats, callback) {
  var url = this._statsUrl(field, stats)
  this.request(url, function (err, json) {
    if (err || json.error) {
      if (!json) json = {error: {}}
      var error = new Error('Request for statistics failed')
      error.timestamp = new Date()
      error.code = json.error.code || 500
      error.body = err || json.error
      error.url = url
      return callback(error)
    }
    callback(null, json)
  })
}

/**
 * Gets the feature service info
 * @param {function} callback - called when the service info comes back
 */
FeatureService.prototype.info = function (callback) {
  if (typeof callback === 'undefined') return this._info
  if (this._info) return callback(null, this._info)
  var url = this.server + '?f=json'
  this.request(url, function (err, json) {
    /**
     * returns error on three conditions:
     * 1. err is present
     * 2. missing response json
     * 3. error in response json
     */
    if (err || !json || json.error) {
      if (!json) json = {error: {}}
      var error = new Error('Request for service information failed')
      error.timestamp = new Date()
      error.url = url
      error.code = json.error.code || 500
      error.body = json.error

      return callback(error)
    }
    this._info = json
    json.url = url
    callback(null, json)
  })
}

/**
 * Gets the feature service layer info
 * @param {function} callback - called when the layer info comes back
 */
FeatureService.prototype.layerInfo = function (callback) {
  // used saved version if available
  if (typeof callback === 'undefined') return this._layerInfo
  if (this._layerInfo) return callback(null, this._layerInfo)

  var url = this.server + '/' + this.layer + '?f=json'

  this.request(url, function (err, json) {
    /**
     * returns error on three conditions:
     * 1. err is present
     * 2. missing response json
     * 3. error in response json
     */
    if (err || !json || json.error) {
      if (!json) json = {error: {}}
      var error = new Error('Request for layer information failed')
      error.timestamp = new Date()
      error.url = url
      error.code = json.error.code || 500
      error.body = json.error

      return callback(error)
    }

    json.url = url
    callback(null, json)
  })
}

/**
 * Gets the objectID field from the service info
 @param {object} info the feature layer metadata
 @returns {string} service's object id field
*/
FeatureService.prototype.getObjectIdField = function (info) {
  var oid
  if (!info.fields) return false
  if (info.objectIdField) return info.objectIdField
  info.fields.some(function (field) {
    if (field.type === 'esriFieldTypeOID') {
      oid = field.name
      return true
    }
  })

  return oid
}

/**
 * Gets the feature service object ids for pagination
 * @param {object} callback - called when the service info comes back
 */
FeatureService.prototype.layerIds = function (callback) {
  var url = this.server + '/' + this.layer + '/query?where=1=1&returnIdsOnly=true&f=json'
  this.request(url, function (err, json) {
    if (err || !json.objectIds) {
      if (!json) json = {error: {}}
      var error = new Error('Request for object IDs failed')
      error.timestamp = new Date()
      error.code = json.error.code || 500
      error.url = url
      error.body = err || json.error

      return callback(error)
    }
    // TODO: is this really necessary
    json.objectIds.sort(function (a, b) { return a - b })
    callback(null, json.objectIds)
  })
}

/**
 * Count of every single feature in the service
 * @param {object} callback - called when the service info comes back
 */
FeatureService.prototype.featureCount = function (callback) {
  var countUrl = this.server + '/' + (this.layer || 0)
  countUrl += '/query?where=1=1&returnCountOnly=true&f=json'

  this.request(countUrl, function (err, json) {
    if (err || json.error) {
      // init empty json error so we can handle building the error in one logic stream
      if (!json) json = {error: {}}
      var error = new Error('Request for feature count failed')
      error.timestamp = new Date()
      error.code = json.error.code || 500
      error.url = countUrl
      error.body = err || json.error

      return callback(error)
    }

    callback(null, json)
  })
}

/**
 * Gets and derives layer metadata from two sources
 * @param {function} callback - called with an error or a metadata object
 */
FeatureService.prototype.metadata = function (callback) {
  if (typeof callback === 'undefined') return this._metadata
  if (this._metadata) return callback(null, this._metadata)

  this.layerInfo(function (err, layer) {
    if (err) {
      err.message = 'Unable to get layer metadata: ' + err.message
      return callback(err)
    }
    this._layerInfo = layer
    var oid = this.getObjectIdField(layer)
    var size = layer.maxRecordCount

    // TODO flatten this
    var metadata = {layer: layer, oid: oid, size: size}

    // 10.0 servers don't support count requests
    // they also do not show current version on the layer
    if (!layer.currentVersion) return callback(null, metadata)

    this.featureCount(function (err, json) {
      if (err) return callback(err)
      if (json.count < 1) return callback(new Error('Service returned count of 0'))
      metadata.count = json.count
      this._metadata = metadata
      callback(null, metadata)
    })
  }.bind(this))
}

/**
 * Build an array pages that will cover every feature in the service
 * @param {function} callback - called when the service info comes back
 */
FeatureService.prototype.pages = function (callback) {
  this.metadata(function (err, meta) {
    if (err) return callback(err)
    if (meta.count < meta.layer.maxRecordCount && meta.count < this.options.size) return callback(null, singlePage(this.server, this.layer, this.outSR))
    this.concurrency = this.options.concurrency || Utils.setConcurrency(this.hosted, meta.layer.geometryType)
    this.maxConcurrency = this.concurrency
    this.pageQueue.concurrency = this.concurrency
    var size = Math.min(parseInt(meta.size, 10), 1000) || 1000
    // restrict page size to the passed in maximum
    if (size > 5000) size = this.options.size

    var layer = meta.layer
    var nPages = Math.ceil(meta.count / size)

    // if the service supports paging, we can use offset to build pages
    var canPage = layer.advancedQueryCapabilities && layer.advancedQueryCapabilities.supportsPagination
    if (canPage) {
      // Test pagination by requesting features from second page - supportsPagination has not been reliable
      var url = this.server + '/0/query?f=json&where=1=1&outFields=*&resultOffset=' + size + '&resultRecordCount=' + size
      this.request(url, function (err, json) {
        if (err) return callback(err)
        // If json features were returned, assume pagination works; otherwise use paging alternatives
        if (json && json.features && json.features.length > 0) {
          return callback(null, this._offsetPages(nPages, size))
        } else {
          this.pagingAlternatives(meta, size, callback)
        }
      }.bind(this))
    } else {
      this.pagingAlternatives(meta, size, callback)
    }
  }.bind(this))
}

/**
 * Build pages when pagination is not supported
 * @param {function} callback - called when the service info comes back
 */
FeatureService.prototype.pagingAlternatives = function (meta, size, callback) {
  if (!meta.oid) return callback(new Error('ObjectID type field not found, unable to page'))
  this.options.objectIdField = meta.oid
  // if the service supports statistics, we can request the maximum and minimum id to build pages
  if (meta.layer.supportsStatistics) {
    this.getObjectIdRange(meta.oid, function (err, stats) {
    // if this worked then we can pagination using where clauses
      if (!err) return callback(null, this._rangePages(stats, size))
      // if it failed, try to request all the ids and split them into pages
      this.layerIds(function (err, ids) {
        // either this works or we give up
        if (err) return callback(err)
        return callback(null, this._idPages(ids, size))
      }.bind(this))
    }.bind(this))
  } else {
    // this is the last thing we can try
    this.layerIds(function (err, ids) {
      if (err) return callback(err)
      callback(null, this._idPages(ids, size))
    }.bind(this))
  }
}

/**
 * Handle feature server request where total number of features can be acquired in a single page
 * @param {*} server
 * @param {*} layer
 * @param {integer} outSR - wkid of output spatial reference
 */
function singlePage (server, layer, outSR = 4326) {
  return [{req: [server, '/', layer, '/query?where=1=1&returnGeometry=true&returnZ=true&outFields=*&outSR=' + outSR + '&f=json'].join('')}]
}

/**
 * Get the max and min object id
 * @param {object} meta - layer metadata, holds information needed to request oid stats
 * @param {function} callback - returns with an error or objectID stats
 */
FeatureService.prototype.getObjectIdRange = function (oidField, callback) {
  this.statistics(oidField, ['min', 'max'], function (err, statResponse) {
    // TODO this is handled elsewhere now so move it
    if (err) return callback(err)
    var eMsg = 'Response from statistics was invalid'
    var stats
    try {
      stats = findMinAndMax(statResponse)
    } catch (e) {
      return callback(new Error(eMsg))
    }
    if (!stats.min > 0 && !stats.max > 0) return callback(new Error(eMsg))
    callback(null, stats)
  })
}

function findMinAndMax (statResponse) {
  var attributes = statResponse.features[0].attributes
  var values = Object.keys(attributes).map(function (key) {
    return attributes[key]
  })
  var minMax = {}
  minMax.min = values[0] < values[1] ? values[0] : values[1]
  minMax.max = values[1] > values[0] ? values[1] : values[0]
  return minMax
}

/**
 * build result Offset based page requests
 * these pages use Server's built in paging via resultOffset and resultRecordCount
 * @param {integer} pages - the number of pages we'll create
 * @param {integer} size - the max number of features per page
 * @returns {object} reqs - contains all the pages for extracting features
 */
FeatureService.prototype._offsetPages = function (pages, size) {
  var reqs = []
  var resultOffset
  var url = this.server

  for (var i = 0; i < pages; i++) {
    resultOffset = i * size
    var pageUrl = url + '/' + this.layer + '/query?outSR=' + this.outSR + '&f=json&outFields=*&where=1=1'
    if (pages === 1) return [{req: pageUrl + '&geometry=&returnGeometry=true&returnZ=true&geometryPrecision='}]
    pageUrl += '&resultOffset=' + resultOffset
    pageUrl += '&resultRecordCount=' + size
    pageUrl += '&geometry=&returnGeometry=true&returnZ=true&geometryPrecision='
    reqs.push({req: pageUrl})
  }

  return reqs
}

/**
 * build `id` query based page requests
 * these pages use object ids in URLs directly
 * @param {array} ids - an array of each object id in the service
 * @param {integer} size - the max record count for each page
 * @returns {object} reqs - contains all the pages for extracting features
 */
FeatureService.prototype._idPages = function (ids, size) {
  var reqs = []
  var oidField = this.options.objectIdField || 'objectId'
  var pages = (ids.length / size)

  for (var i = 0; i < pages + 1; i++) {
    var pageIds = ids.splice(0, size)
    if (pageIds.length) {
      var pageMin = pageIds[0]
      var pageMax = pageIds.pop()
      var where = [oidField, ' >= ', pageMin, ' AND ', oidField, '<=', pageMax].join('')
      var pageUrl = this.server + '/' + (this.layer) + '/query?outSR=' + this.outSR + '&where=' + where + '&f=json&outFields=*'
      pageUrl += '&geometry=&returnGeometry=true&returnZ=true&geometryPrecision=10'
      reqs.push({req: pageUrl})
    }
  }

  return reqs
}

/**
 * build object id query based page requests
 * these pages use object ids in where clauses via < and >
 * you could call this objectId queries
 * @param {object} stats - contains the max and min object id
 * @param {integer} size - the size of records to include in each page
 * @returns {object} reqs - contains all the pages for extracting features
 */
FeatureService.prototype._rangePages = function (stats, size) {
  var reqs = []
  var pageUrl
  var pageMax
  var pageMin
  var where
  var objId = this.options.objectIdField

  var url = this.server
  var pages = Math.max((stats.max === size) ? stats.max : Math.ceil((stats.max - stats.min) / size), 1)

  for (var i = 0; i < pages; i++) {
    // there is a bug in server where queries fail if the max value queried is higher than the actual max
    // so if this is the last page, then set the max to be the maxOID
    if (i === pages - 1) {
      pageMax = stats.max
    } else {
      pageMax = stats.min + (size * (i + 1)) - 1
    }
    pageMin = stats.min + (size * i)
    where = [objId, '>=', pageMin, '+AND+', objId, '<=', pageMax].join('')
    pageUrl = url + '/' + (this.layer || 0) + '/query?outSR=' + this.outSR + '&where=' + where + '&f=json&outFields=*'
    pageUrl += '&geometry=&returnGeometry=true&returnZ=true&geometryPrecision='
    reqs.push({req: pageUrl})
  }

  return reqs
}

/**
 * Requests a page of features
 * @param {object} task - a task object with a "req" property
 * @param {function} callback
 */
FeatureService.prototype._requestFeatures = function (task, cb) {
  var self = this

  this.request(task.req, function (err, json) {
    if (err) return self._catchErrors(task, err, task.req, cb)
    if (!json || json.error) {
      if (!json) json = {error: {}}
      var error = new Error('Request for a page of features failed')
      error.timestamp = new Date()
      error.body = json.error
      error.code = json.error.code || 500
      return self._catchErrors(task, error, task.req, cb)
    }
    self._throttleQueue()
    cb(null, json)
  })
}

/* Catches an errors during paging and handles retry logic
 * @param {object} task - the currently executing job
 * @param {object} e - the error in application logic or from a failed request to a server
 * @param {string} url - the url of the last request for pages
 * @param {function} cb - callback passed through to the abort paging function
 */
FeatureService.prototype._catchErrors = function (task, error, url, cb) {
  this._throttleQueue(error)
  // be defensive in case there was no json payload
  error.body = error.body || {}
  // set the error code from the json payload if the error doesn't have one already
  if (!error.code) error.code = error.body.code
  error.url = url
  if (task.retry && task.retry === 3) return this._abortPaging(error, cb)
  // initiate the count or increment it
  if (!task.retry) {
    task.retry = 1
  } else {
    task.retry++
  }

  this.log('info', 'Re-requesting page ' + task.req + ' attempt ' + task.retry)

  setTimeout(function () {
    this._requestFeatures(task, cb)
  }.bind(this), task.retry * this.options.backoff)
}

FeatureService.prototype._throttleQueue = function (fail) {
  if (fail) this.concurrency -= 0.5
  else this.concurrency += 0.1

  if (this.concurrency > this.maxConcurrency) this.concurrency = this.maxConcurrency

  this.pageQueue.concurrency = this.concurrency >= 1 ? Math.floor(this.concurrency) : 1
}

/**
 * Aborts the request queue by emptying all queued up tasks
 * @param {object} error - error payload to send back to the original requestor
 * @param {function} callback - calls back with the error payload
 */
FeatureService.prototype._abortPaging = function (error, callback) {
  this.pageQueue.kill()
  error.message = 'Paging aborted: ' + error.message
  callback(error)
}

module.exports = FeatureService
